File: rxjavademo/src/main/java/com/soda/rxjavademo/MainActivity.java
package com.soda.rxjavademo;

import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.View;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class MainActivity extends AppCompatActivity {


    private static final String TAG = "rxjavademo.MainActivity";
    Observable<String> mSwitcher;
    Observer<String> mLight;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //标准方式生成被观察者(观察源)
        mSwitcher = Observable.create(new ObservableOnSubscribe<String>(){

            @Override
            public void subscribe(ObservableEmitter<String> emitter) throws Exception {
                emitter.onNext("On");
                emitter.onNext("Off");
                emitter.onComplete(); //下游接收一次compete/error事件之后就不再执行onNext方法
                //onComplete方法在下游只会响应一次,所以多次调用并无意义
//                emitter.onError(new RuntimeException("test exception"));
                emitter.onNext("On");
                emitter.onNext("off");
            }
        });
        //创建被观察者的简便方式,使用just,还可以使用数组加fromArray操作符
//        mSwitcher = Observable.just("on", "off", "on");
        mLight = new Observer<String>() {

            @Override
            public void onSubscribe(Disposable d) {
                Log.d(TAG, "onSubscribe");
            }

            @Override
            public void onNext(String s) {
                Log.d(TAG, "onNext : " + s);
            }

            @Override
            public void onError(Throwable t) {
                Log.d(TAG, "onError : " + t.getMessage());
            }

            @Override
            public void onComplete() {
                Log.d(TAG, "onComplete");
            }
        };
    }

    public void switchLight(View view) {
        mSwitcher.subscribe(mLight);
    }

    public void useFlowable(View view) {
        Intent intent = new Intent(this, FlowableActivity.class);
        startActivity(intent);
    }

}

File: arcfacetest/src/main/java/com/palmax/arcfacetest/MainActivity.kt
package com.palmax.arcfacetest

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.widget.Toast
import com.arcsoft.face.ErrorInfo
import com.palmax.arcfacelib.ArcFaceManager
import com.palmax.arcfacelib.FaceRepository
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

class MainActivity : AppCompatActivity() {

    private val NEEDED_PERMISSIONS = arrayOf(Manifest.permission.READ_PHONE_STATE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    private val ACTION_REQUEST_PERMISSIONS = 0x001

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if (!checkPermissions(NEEDED_PERMISSIONS)) {
            ActivityCompat.requestPermissions(this, NEEDED_PERMISSIONS, ACTION_REQUEST_PERMISSIONS)
        } else {
            ArcFaceManager.init(this)
        }
    }

    /**
     * 激活引擎
     *
     */
    fun activeEngine(v:View) {
        Observable.create(ObservableOnSubscribe<Int> {
            it.onNext(FaceRepository.activeEngine(this))
            it.onComplete()
        }).subscribe(
                object : Observer<Int> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(activeCode: Int) {
                        when (activeCode) {
                            ErrorInfo.MOK -> showToast(getString(R.string.active_success))
                            ErrorInfo.MERR_ASF_ALREADY_ACTIVATED -> showToast(getString(R.string.already_activated))
                            else -> showToast(getString(R.string.active_failed, activeCode))
                        }
                    }

                    override fun onError(e: Throwable) {
                        showToast("active error")
                    }

                    override fun onComplete() {

                    }
                }
        )

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == ACTION_REQUEST_PERMISSIONS) {
            var isAllGranted = true
            for (grantResult in grantResults) {
                isAllGranted = isAllGranted and (grantResult == PackageManager.PERMISSION_GRANTED)
            }
            if (isAllGranted) {
                ArcFaceManager.init(this)
            } else {
                showToast(getString(R.string.permission_denied))
            }
        }
    }

    private fun checkPermissions(neededPermissions: Array<String>?): Boolean {
        if (neededPermissions == null || neededPermissions.isEmpty()) {
            return true
        }
        var allGranted = true
        for (neededPermission in neededPermissions) {
            allGranted = allGranted and (ContextCompat.checkSelfPermission(this, neededPermission) == PackageManager.PERMISSION_GRANTED)
        }
        return allGranted
    }

    private fun showToast(s: String) {
       Toast.makeText(this, s, Toast.LENGTH_SHORT).show()
    }

}
